### ISSUE TITLE
Number literals handle overflow and underflow inconsistently

### ISSUE THREAD

**Issue by vpetrovykh** [edgedb/edgedb#1402]
Edgedb version:  EdgeDB 1.0-alpha.3+dev.198.g9e8c0613
REPL version: edgedb-cli 0.1.0

I was a little surprised by the way finite-bit numbers are handled. The overflow and underflow doesn't happen consistently.

Consider the following for overflow:
```
victor> SELECT 11111111111111111111111111111111111111111111;
ERROR: EdgeQLSyntaxError: can't parse integer: number too large to fit in target type
```
So far so good, the `int64` overflowed as expected.

Let's see what the float does:
```
victor> SELECT 1e999;
{inf}
victor> SELECT 111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111.0;
{inf}
```
That's peculiar, I was expecting an error, much like with the `int64`. So let's try slightly different literals:
```
victor> SELECT <float64>'inf';
{inf}
victor> SELECT <float64>'1e999';
ERROR: NumericOutOfRangeError: "1e999" is out of range for type std::float64
victor> SELECT <float64>1e999n;
ERROR: NumericOutOfRangeError: "1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000" is out of range for type std::float64
```

Similarly, there's an underflow issue:
```
victor> SELECT 1e-999;
{0}
victor> SELECT <float64>1e-999n;
ERROR: NumericOutOfRangeError: "0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001" is out of range for type std::float64
victor> SELECT <float64>to_json('1e-999');
ERROR: NumericOutOfRangeError: "0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001" is out of range for type std::float64
```

So it appears that overflow and underflow are not handled consistently, which can be a weird foot-gun.

**Comment by tailhook** [comment]
Look like a constant extraction issue. Rust says 'literal out of range' when it's literal in rust code, but goes with `inf` when parsing the value at runtime.

**Comment by tailhook** [comment]
So `+1e999` and `-1e999` is clear how to handle (basically error out if the resulting value is infinity).

But it looks like there is nothing wrong with `1e-999` being `0`. This kind of inaccuracies appear every time you're working with floats. I also don't see any good way to fix that. Should we restrict the way of literal zero could be written? That looks like too error-prone.

**Comment by tailhook** [comment]
So any comments on `1e-999`? I would close this after #1523

**Comment by 1st1** [comment]
> So +1e999 and -1e999 is clear how to handle (basically error out if the resulting value is infinity).

Hm, Python happily consumes both. Is it possible to make them always work and convert to infinity?

**Comment by tailhook** [comment]
Do you mean replacing postgres conversion function? I think it's possible, but would make all str->float conversions slower.

**Comment by 1st1** [comment]
> but would make all str->float conversions slower.

Ah, so `ERROR: NumericOutOfRangeError: "10000000` comes from PostgreSQL, and not from the constant extraction. This changes things. Making float conversions slower isn't an option :/

**Comment by tailhook** [comment]
> > but would make all str->float conversions slower.
> 
> Ah, so `ERROR: NumericOutOfRangeError: "10000000` comes from PostgreSQL, and not from the constant extraction. This changes things. Making float conversions slower isn't an option :/

Yes. I've fixed that. The only thing left, is whether we want `1e-999` be a `0` or an error. While postgres crashes on this. I think it's fine to allow that in edgeql as it's not ambiguous (comparing to `1e+999` which as infinity).

**Comment by 1st1** [comment]
But we ought to be consistent, I think. IMO, if `<float64>'1e-999'` is an error, the `1e-999` literal should be too.

**Comment by tailhook** [comment]
```
edgedb> SELECT <float32>'1e-99';
ERROR: NumericOutOfRangeError: "1e-99" is out of range for type std::float32
edgedb> SELECT <float64>'1e-99';
{0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001}
```
I assume that you want literals be float64 always, right?

**Comment by tailhook** [comment]
Fixed in #1950. But now I remember what bothers me:
```
NumericOutOfRangeError: "1e-999" is out of range for type std::float64
```

Why this is "Out of range"? It's just zero `0`, it's definitely withing the range of float, somewhere between `0.1` and `-0.1`. And I don't have a better error description for this in plain English too.

**Comment by 1st1** [comment]
> I assume that you want literals be float64 always, right?

Yes, IIRC our numeric literals are float64 if they have numbers after `.`.

> Why this is "Out of range"? It's just zero 0, it's definitely within the range of float, somewhere between 0.1 and -0.1. And I don't have a better error description for this in plain English too.

Perhaps we should talk more about loosing precision than about the range. Ultimately `1e-999` is not `0` in mathematical sense and is very different from `1e-99`, it's just we'll lose that difference if we assume it's 0.

So maybe instead of `"1e-999" is out of range for type std::float64` we say `"1e-999" is not representable as std::float64 without losing precision (try using std::decimal)`?

@elprans @vpetrovykh

**Comment by tailhook** [comment]
> > 
> 
> Perhaps we should talk more about loosing precision than about the range. Ultimately `1e-999` is not `0` in mathematical sense and is very different from `1e-99`, it's just we'll lose that difference if we assume it's 0.

There are more cases when might loose precision than just zero. This message would create unjustified expectations.

**Comment by 1st1** [comment]
> generally loosing precision is how floats work by default, that's why we don't use == on them and so on

Yeah, I know. I also kind of agree with "This message would create unjustified expectations.". Let's keep the "out of range" message.
